### PR TITLE
Implemented fraction intrinsic function

### LINKED ISSUE #3400: Implement Fraction intrinsic function
Refer : https://gcc.gnu.org/onlinedocs/gfortran/FRACTION.html

### PR DESCRIPTION
**Pull request by parth121101**
Fix #3400

### CODE REVIEW

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -2260,6 +2261,43 @@ namespace Exp {
...
+
+    static inline ASR::expr_t* instantiate_Fraction(Allocator &al, const Location &loc,
+            SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("");
+        fill_func_arg("x", arg_types[0]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+        /*
+        * r = fraction(x, y)
+        * r = x * 2**(-exp(x))
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2280]:
```suggestion
        * r = x * radix(x)**(-exp(x))
```

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -2260,6 +2261,43 @@ namespace Exp {
...
+        */
+
+       //TODO: Radix for most of the device is 2, so we can use the i2r32(2) instead of args[1]. Fix (find a way to get the radix of the device and use it here)
+        ASR::ttype_t* return_type_real = expr_type(args[0]);
+
+        Vec<ASR::ttype_t*> arg_types_mod; arg_types_mod.reserve(al, 1);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2286]:
```suggestion
        Vec<ASR::ttype_t*> arg_types_exp; arg_types_exp.reserve(al, 1);
```

File: src/libasr/pass/intrinsic_functions.h [hunk 3]
```diff
@@ -2260,6 +2261,43 @@ namespace Exp {
...
+        arg_types_mod.push_back(al, arg_types[0]);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2287]: same comment as above

File: src/libasr/pass/intrinsic_functions.h [hunk 4]
```diff
@@ -2260,6 +2261,43 @@ namespace Exp {
...
+
+        Vec<ASR::call_arg_t> new_args_mod; new_args_mod.reserve(al, 1);
+        ASR::call_arg_t arg1; arg1.loc = loc; arg1.m_value = args[0];
+        new_args_mod.push_back(al, arg1);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2291]: same comment as above

File: integration_tests/intrinsics_138.f90 [hunk 5]
```diff
@@ -0,0 +1,8 @@
+program intrinsics_138
+    real :: x
+    real :: y
+    x = 178.1387e-4
+    y = 1.00
+
+    print *, fraction(x), x * radix(x)**(-exp(x))
+end program
```
Comment (Pranavchiku) [inline integration_tests/intrinsics_138.f90:8]: Please add tests with `if(abs(xx)) error stop` conditions.

File: src/libasr/pass/intrinsic_functions.h [hunk 6]
```diff
@@ -2260,6 +2261,43 @@ namespace Exp {
...
+            ASR::ttype_t* arg_type, Vec<ASR::expr_t*> &args, diag::Diagnostics& /*diag*/) {
+        double value_X = ASR::down_cast<ASR::RealConstant_t>(expr_value(args[0]))->m_r;
+        double result = value_X * std::pow(2, (-1*(std::exp(value_X))));
+        return f(result, arg_type);
+    }
+
+    static inline ASR::expr_t* instantiate_Fraction(Allocator &al, const Location &loc,
+            SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("");
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2275]:
```suggestion
        declare_basic_variables("_lcompilers_fraction_" + type_to_str_python(arg_types[0]));
```

File: src/libasr/pass/intrinsic_functions.h [hunk 7]
```diff
@@ -2260,6 +2261,43 @@ namespace Exp {
...
+
+        Vec<ASR::ttype_t*> arg_types_mod; arg_types_mod.reserve(al, 1);
+        arg_types_mod.push_back(al, arg_types[0]);
+
+        Vec<ASR::call_arg_t> new_args_mod; new_args_mod.reserve(al, 1);
+        ASR::call_arg_t arg1; arg1.loc = loc; arg1.m_value = args[0];
+        new_args_mod.push_back(al, arg1);
+
+        ASR::expr_t* func_call_exp = Exp::instantiate_Exp(al, loc, scope, arg_types_mod, return_type_real, new_args_mod, 0);
+        body.push_back(al, b.Assignment(result, r_tMul(args[0], rPow(i(2, arg_types[0]), func_call_exp, arg_types[0]), arg_types[0])));
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2294]:
```suggestion
        func_call_exp = EXPR(ASR::make_RealUnaryMinus_t(al, loc, func_call_exp, return_type_real, nullptr));
        body.push_back(al, b.Assignment(result, r_tMul(args[0], rPow(i2r(i(2, int32),return_type_real), func_call_exp, arg_types[0]), arg_types[0])));  
```

**Comment by Pranavchiku** [comment]
For the given example, we get different output with gfortran, `fraction(x) = 0.570043862` but `x * radix(x) ** (-exp(x)) = 8.79665837E-03`. The documentation statest `fraction(x) = x * radix(x) ** (-exp(x))`.

Reference link: https://gcc.gnu.org/onlinedocs/gfortran/FRACTION.html

```console
% gfortran ./integration_tests/intrinsics_138.f90 && ./a.out
  0.570043862       8.79665837E-03
```

cc @Thirumalai-Shaktivel @Shaikh-Ubaid

**Comment by HarshitaKalani** [comment]
The documentation states `X * RADIX(X)**(-EXPONENT(X))` and not `X * RADIX(X)**(-EXP(X))`. They're two different functions actually.
EXP: https://gcc.gnu.org/onlinedocs/gfortran/EXP.html
EXPONENT: https://gcc.gnu.org/onlinedocs/gfortran/EXPONENT.html

File: integration_tests/intrinsics_138.f90 [hunk 8]
```diff
@@ -0,0 +1,8 @@
+program intrinsics_138
+    real :: x
+    real :: y
+    x = 178.1387e-4
+    y = 1.00
+
+    print *, fraction(x), x * radix(x)**(-exp(x))
```
Comment (HarshitaKalani) [inline integration_tests/intrinsics_138.f90:7]:
```suggestion
    print *, fraction(x), x * radix(x)**(-exponent(x))
```

**Comment by parth121101** [comment]
> The documentation states `X * RADIX(X)**(-EXPONENT(X))` and not `X * RADIX(X)**(-EXP(X))`. They're two different functions actually. EXP: https://gcc.gnu.org/onlinedocs/gfortran/EXP.html EXPONENT: https://gcc.gnu.org/onlinedocs/gfortran/EXPONENT.html

I think the exponent function is missing. So I need to first add the exponent function. So till that time, I think this PR needs to be kept at a halt. What's your suggestion @Pranavchiku @HarshitaKalani ?

File: src/libasr/intrinsic_func_registry_util_gen.py [hunk 9]
```diff
@@ -230,6 +230,12 @@
             "ret_type_arg_idx": 0
         },
     ],
+    "Fraction": [
+        {
+            "args": [("real",)],
+            "ret_type_arg_idx": 0
+        },
+    ],
```
Comment (certik) [inline src/libasr/intrinsic_func_registry_util_gen.py:238]: I think this function should go to `compile_time_only_fn`.

**Comment by certik** [comment]
@parth121101 yes, first implement `exponent`, then you can implement `fraction`. I think they are both compile time only functions, that depend on a type, not the value.

**Comment by parth121101** [comment]
Closing in favor of #3583

**Pull request closed without merging**